server/pypi/make.sh

```diff
@@ -1,57 +1,285 @@
+PYTHON_EXACT_VERSIONS["3.8"]="3.8.16"
+PYTHON_EXACT_VERSIONS["3.9"]="3.9.16"
+PYTHON_EXACT_VERSIONS["3.10"]="3.10.9"
+PYTHON_EXACT_VERSIONS["3.11"]="3.11.0"
+BUILD_DEPS="1"
+BUILD_ENVS="1"
+YEAR="2021"
+
+# dependencies to build
+
+DEPENDENCIES="
+    chaquopy-freetype \
+    chaquopy-libjpeg \
+    chaquopy-libogg \
+    chaquopy-libpng \
+    chaquopy-libxml2 \
+    chaquopy-libiconv \
+    chaquopy-curl \
+    chaquopy-ta-lib \
+    chaquopy-zbar \
+    "
+
+# TODO: to investigate if we need those dependencies
+
+# chaquopy-crc32c
+# chaquopy-flac
+# chaquopy-geos
+# chaquopy-hdf5
+# chaquopy-libcxx
+# chaquopy-libffi
+# chaquopy-libgfortran
+# chaquopy-libomp
+# chaquopy-libraw
+# chaquopy-libsndfile
+# chaquopy-libvorbis
+# chaquopy-libxslt
+# chaquopy-libzmq
+# chaquopy-llvm
+# chaquopy-openblas
+# chaquopy-secp256k1
+
+# packages to build
+
+PACKAGES="
+    cffi \
+    numpy \
+    aiohttp \
+    argon2-cffi \
+    backports-zoneinfo \
+    bcrypt \
+    bitarray \
+    brotli \
+    cryptography \
+    cymem \
+    cytoolz \
+    editdistance \
+    ephem \
+    frozenlist \
+    gensim \
+    greenlet \
+    kiwisolver \
+    lru-dict \
+    matplotlib \
+    multidict \
+    murmurhash \
+    netifaces \
+    pandas \
+    pillow \
+    preshed \
+    pycrypto \
+    pycurl \
+    pynacl \
+    pysha3 \
+    pywavelets \
+    pyzbar \
+    regex \
+    ruamel-yaml-clib \
+    scandir \
+    spectrum \
+    srsly \
+    statsmodels \
+    twisted \
+    typed-ast \
+    ujson \
+    wordcloud \
+    yarl \
+    zstandard \
+    "
+# TODO: to investigate if we need to build those packages for iOS
+
+# blis
+# cmake-example
+# coincurve
+# cvxopt
+# dlib
+# gevent
+# google-crc32c
+# grpcio
+# h5py
+# llvmlite
+# lxml
+# numba
+# opencv-contrib-python
+# opencv-contrib-python-headless
+# opencv-python
+# opencv-python-headless
+# psutil
+# pycares
+# pycryptodome
+# pycryptodomex
+# python-example
+# pyzmq
+# rawpy
+# rpi-gpio
+# scikit-image
+# scikit-learn
+# scipy
+# sentencepiece
+# shapely
+# soundfile
+# spacy
+# ta-lib
+# tensorflow
+# tflite-runtime
+# thinc
+# tokenizers
+# torch
+# torchvision
+# xgboost
+
+# parse command line
+
+POSITIONAL=()
+while [[ $# -gt 0 ]] ; do
+  key="${1}"
+
+  case ${key} in
+    --help)
+      echo "Usage: make.sh options
+
+      options
+              --package package
+              --skip-deps
+              --skip-envs
+              --year year"
+      exit 0
+      ;;
+    --package)
+      PACKAGES=${2}
+      shift
+      shift
+      ;;
+    --skip-deps)
+      BUILD_DEPS="0"
+      shift
+      ;;
+    --skip-envs)
+      BUILD_ENVS="0"
+      shift
+      ;;
+    --year)
+      YEAR=${2}
+      shift
+      shift
+      ;;
+    *)
+      POSITIONAL+=("$1") # save it in an array for later
+      shift
+    ;;
+  esac
+done
+
+set -- "${POSITIONAL[@]}" # restore positional parameters
+
+# setup clean conda environments and toolchains for each Python versions
```

Comment:
Definitively inside a GitHub action this would be a matrix running the 4 Python versions in parallel without a conda environment.

I'm just trying to have a fixed environment here so we can all work on the same base to fix issues. It doesn't have to be conda at all but it's like easy for us to put each machine in the exact same configuration for now until all packages can build.

I could move it to a second script but the main script must still keep knowledge of it if we keep the loop on the Python version.

I would wait for the harder steps to be completed and take some steps back before we tamper too much with things that may change. YOu said you wanted to keep it common with the main Android repo so we also need to consider this in the mix.

Reply:
Having a stable base definitely makes sense (and conda makes at least some sense as a base); however, I'd argue there's a difference between agreeing on a stable base, and baking that in as a requirements of the script's operation. At the very least, I'm a little wary of introducing a hard requirement on Anaconda because of the conflict of interest I have as a result of being employed by Anaconda. An external requirement of "we're going to use Conda for CI purposes, but it will run in any Python environment" is different to an internal requirement of "this script requires that you have conda installed to run it at all".

Your comments you seem to imply that this script could be executed in parallel as-is - I'm not sure I'm see how. What would be the mechanism for this?

Reply:
That script would need serious rewriting to be used inside a parallel CI pipeline. There are two things to consider.

Initial build of a package is probably done manually and will loop on many package versions. This could also be automated in CI with smart commit tags. We could do as part of a commit something like `#build-all-versions` and the loop will be on the CI side. But we need to monitor the CI for failure and some part may fail just because Python 3.11 is not well supported yet and a human will quickly ignore those failures.

Once the initial build is done the CI should take over on monitoring new tag, build only the required new part and deliver the wheel directly.

Reply:
Ok - but if the end goal is a CI configuration that can build a single version for a single Python, wouldn't that support the idea of having an external "build everything" script that does this sort of iteration, but keep the core build script single target with an externally assumed environment?

Reply:
Yes we agree on this 100%. To me this is `make.sh`. The GitHub action will not need `make.sh` and will directly call `build-wheel.py`.

Let's try to solve a few issues. How do we remove the version from the yaml so it's not linked to a fixed version while at the same time we keep a patch scheme flexible?

I will put in place a prototype a a CI action monitoring remote repo so you better understand.